Library/Homebrew/cmd/--env.rb

```diff
@@ -3,20 +3,36 @@
 
 require "extend/ENV"
 require "build_environment"
+require "utils/shell"
 
 module Homebrew
+  def explicit_shell
+    shell = ARGV.value("shell")
+    return :noshell if shell == "none"
+    Utils::Shell.path_to_shell(shell) if shell
+  end
+
   def __env
     ENV.activate_extensions!
     ENV.deps = ARGV.formulae if superenv?
     ENV.setup_build_environment
     ENV.universal_binary if ARGV.build_universal?
 
-    if $stdout.tty?
+    # explicit shell takes precedence over everything
+    # then we detect if the output is a tty
+    # then we check what kind of shell the parent process is
+    # finally we consult the environment variable
+    shell = explicit_shell
+    shell ||= :noshell if $stdout.tty?
```

Comment:
I think you can make shell `nil` in this case and use that everywhere you should have used `:noshell`.